File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.treeSt = []
        curr = root
        while curr:
            self.treeSt.append(curr)
            curr = curr.left

    def next(self) -> int:
        curr = self.treeSt.pop()
        if curr.right:
            nxt = curr.right
            while nxt:
                self.treeSt.append(nxt)
                nxt = nxt.left

        return curr.val

    def hasNext(self) -> bool:
        return self.treeSt


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> stack;
    
    public BSTIterator(TreeNode node) {
        stack = new Stack<>();
        TreeNode root = node;
        updateStack(root);
    }
    
    public int next() {
        TreeNode currNode = stack.pop();
        updateStack(currNode.right);
        return currNode.val;
    }
    
    public boolean hasNext() {
        return !stack.isEmpty();
    }
    
    private void updateStack(TreeNode root) {
        while (root != null) {
            stack.add(root);
            root = root.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */